.travis.yml

```diff
@@ -7,26 +7,126 @@ before_install:
 sudo: false
 
 matrix:
-    include:        
+    include:
         - python: "3.5"
         - python: "3.6"
         - python: "3.7-dev"
-        - env:
-            - BUILD_DOCKER=1
-            # Variable to add to publish the Docker image:
-            # * DOCKER_USERNAME
-            # * DOCKER_PASSWORD, to be encrypted, use `travis encrypt DOCKER_PASSWORD=<password>`
+        - stage: build_docker
+          name: amd64 docker build
           services:
             - docker
           before_install:
             - true
           install:
             - true
+          after_success:
+            - true
           script:
-            - docker build --tag=the-paperless-project/paperless .
+            - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+            - docker build -f Dockerfile --tag=${DOCKER_REPO}:${TRAVIS_COMMIT}-amd64 .
+            - docker push ${DOCKER_REPO}:${TRAVIS_COMMIT}-amd64
+          on:
+            condition: '"${BUILD_DOCKER}" = 1'
+        - stage: build_docker
+          name: arm64v8 docker build
+          services:
+            - docker
+          before_install:
+            - true
+          install:
+            - true
           after_success:
             - true
-
+          script:
+            - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+            # travis_wait 60 tells travis to wait for up to 60 minutes - default is 20, which is too short
+            - travis_wait 60 docker build -f Dockerfile --tag=${DOCKER_REPO}:${TRAVIS_COMMIT}-arm64v8 .
+            - docker push ${DOCKER_REPO}:${TRAVIS_COMMIT}-arm64v8
+          arch: arm64
+          on:
+            condition: '"${BUILD_DOCKER}" = 1'
+        - stage: build_docker
+          name: arm32v7 docker build
+          services:
+            - docker
+          before_install:
+            - true
+          install:
+            - true
+          after_success:
+            - true
+          script:
+            - echo "$DOCKER_PASSWORD" | docker login -u "$DOCKER_USERNAME" --password-stdin
+            # register binfmt stuff for qemu-static binaries so we can use userland-emulation
+            - docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+            # replace the multi-arch reference with a specific, arm32v7 version. else docker will use the platform specific one,
+            # which is amd64.
+            - sed -i 's/FROM alpine:3.10/FROM alpine@sha256:2632d6288d34d7175021683f6e363fa7c0fa8866a565eb285e36e3b856545e82/g' Dockerfile
```

Comment:
I'm not sure that using the sha is good for maintenance...

Reply:
indeed using the sha directly was not my preferred solution. the underlying problem I'm trying to solve is, that the alpine:3.10 image is a multi-arch image. when using qemu in docker to emulate the arm32v7 cpu architecture on x86_64, docker still pulls the x86_64 image instead of the arm32v7 one. by specifying the image directly and not the manifest (which points to the individual images depending on the operating system and architecture) we can circumvent that problem.

I tried telling docker which architecture I would like it to build for, but failed finding any usable documentation and only vague hints towards buildx, which seems to be a complete mess :shrug: 

So I'm open to suggestions - but I don't see how to solve this without tagging a specific image.

Reply:
Ok, thanks for your explain...